### DISCUSSION TITLE
transient--... already defined as something else than a generic function

### DISCUSSION THREAD

**Discussion by starenka** [magit/magit#5180] [Support] [answered]
i get this while trying to invoke `magit-status`

```
Debugger entered--Lisp error: (error "transient--init-suffix-key is already defined as s...")
  error("%s is already defined as something else than a generic function" transient--init-suffix-key)
  cl-generic-ensure-function(transient--init-suffix-key)
  cl-generic-define-method(transient--init-suffix-key nil ((obj transient-suffix)) nil #f(compiled-function (obj) #<bytecode 0x178f025ef48a9c58>))
  byte-code("\300\301\302\303\302\304%\210\300\301\302\305\306\307%\207" [cl-generic-define-method transient--init-suffix-key nil ((obj transient-suffix)) #f(compiled-function (obj) #<bytecode 0x178f025ef48a9c58>) ((obj transient-argument)) curried #f(compiled-function (cl--nm) #<bytecode 0x14293ab374ff7107>)] 6)
  require(transient)
  load-with-code-conversion("/home/starenka/.emacs.d/elpa/magit-20240721.1509/m..." "/home/starenka/.emacs.d/elpa/magit-20240721.1509/m..." nil t)
  require(magit-mode)
  load-with-code-conversion("/home/starenka/.emacs.d/elpa/magit-20240721.1509/m..." "/home/starenka/.emacs.d/elpa/magit-20240721.1509/m..." nil t)
  require(magit-core)
  load-with-code-conversion("/home/starenka/.emacs.d/elpa/magit-20240721.1509/m..." "/home/starenka/.emacs.d/elpa/magit-20240721.1509/m..." nil t)
  require(magit)
  load-with-code-conversion("/home/starenka/.emacs.d/elpa/magit-20240721.1509/m..." "/home/starenka/.emacs.d/elpa/magit-20240721.1509/m..." nil t)
  command-execute(magit-status)
```

i can't even get `magit-version` for same error, my magit version is `20240721.1509`

i'm running `GNU Emacs 29.4 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.42, cairo version 1.18.0)
 of 2024-06-30, modified by Debian`

**Comment by tarsius** [comment]
> transient--init-suffix-key is already defined as something else than a generic function

This is an "impossible" error, this was never defined as something other than a generic function. A few users had similar issues a few months ago, but we couldn't figure out what was going on (and I cannot find the discussions about it now). 

As always with "impossible" errors, the solution is to start over from a clean slate.

1. Uninstall `magit` and `transient` (and other packages depending on the latter). **Edit: Also uninstall `compat`.**
2. Exit Emacs.
3. Restart Emacs.
4. Reinstall the packages.

**Reply by starenka** [reply]
i did that before reporting. the transient package was the key tho. i forget about that one. sorry!

**Reply by mattsears** [reply]
I'm receiving the same error after upgrading to the latest version of Magit. I've tried uninstalling, restarting, and reinstalling and still receive the same error message when invoking `magit-status`. Currently stuck with magit-20240704.2248 the last version that worked.

**Reply by tarsius** [reply]
All I know is that there is nothing wrong in Magit as distributed and that there must be something wrong in the files that are installed on your machine. Do more extreme variants of the re-installation dance. Remove packages while Emacs isn't running, by deleting directories. Also remove native compiled files. Remove all packages (but first make sure you know how to reinstall them). Use the latest Emacs release. Not an older release and also not a development version. After changing the Emacs version, you must always reinstall everything from scratch.

**Reply by mattsears** [reply]
Cool, did all that ☝🏻 and the latest version of Magit works, but `transient-20240629.1508` is the last version that worked for me. Anything newer, I get the error message 🤷🏻 

Restoring `defmacro static-if` (https://github.com/magit/transient/commit/ed732cb35f6a6753930df10a0f66a6e4efd29e6d) on `transient-20240713.2102` locally fixed the issue for me.

**Reply by tarsius** [reply]
I don't see how that could be related, but since we have no other leads let's investigate that some more.

Is `compat-30.0.0.0` installed? Does it get installed before `transient` and `magit`?

Provided that `compat` is installed (early enough), it should make no difference whether `transient` also (re-)defines it or not, the two versions should be the same. However it is possible that first `compat` defines it, then another package re-defines it but that implementation is broken. Then it would make a difference whether `transient` later re-re-defines it to the proper implementation.

Grep all installed packages for `defmacro static-if`. I grepped all packages on the Emacsmirror and found one broken implementation in the `apel` package.

**Reply by mattsears** [reply]
My grep results show `swiper-20240520.120` (https://github.com/abo-abo/swiper/blob/2a25a6fb5b081cb141c5eccac8ee58ab1feeb747/swiper.el#L43)  and `ivy-20240524.1139` (https://github.com/abo-abo/swiper/blob/2a25a6fb5b081cb141c5eccac8ee58ab1feeb747/ivy.el#L53) packages calling `defmacro static-if` locally. 

`compat-30.0.0.0` is installed as well. I'm using Cask for package management and not sure the order in which it installs packages.

**Reply by tarsius** [reply]
> I'm using Cask for package management

Did you do that:

> + Why does this +shit+ never work? :: Bitrot and weak hacks, primarily.  Always try =rm -rf ~/.emacs.d/.cask= since we're liable to make backwards-incompatible changes.

**But first**, if you haven't done so yet, then please post the output of `tree -a elpa .cask` first. I suspect old versions of the relevant packages are laying around in `.cask`.

I believe to remember that one of those backward-incompatible changes was to make Cask unsuitable for package management. More precisely I remember a user complaining about a change that broke their use of Cask as a package manager and the then new maintainer replying that Cask should not be used as a package manager (only as a package development tool) and that the changes won't be reverted. This recollection might not be 100% accurate. I couldn't find that issue, but see the somewhat related https://github.com/cask/cask/issues/581.

Honestly, I always felt that what Cask was too much inspired by tools for languages such as Ruby and Python, but that that approach is completely unsuitable for Emacs, where all installed packages run in the same `emacs` process, which is very different from running, e.g., two python web applications in two separate processes. In Emacs two packages cannot use two different versions of a shared dependency. It is my impression that Cask is optimized for this "impossible" use-case.

I suspect that as a result of this fundamental mismatch, updating some package does not actually mean that all packages that depend on that package get compiled against that version. Some package might get compiled against an old version laying around in `.cask` directories. I have not verified this, it's just a suspicion.

**Reply by mattsears** [reply]
> Why does this +shit+ never work? 

That hasn't been my experience with Cask. It's served me quite well keeping packages up-to-date over the decade and rarely run into issues. I'm not planning on switching any time soon. 

It's all good though - I will just keep the restored `static-if` macro locally in `transient.el` for now.  Thanks!

**Reply by tarsius** [reply]
>> Why does this +shit+ never work?

> That hasn't been my experience with Cask.

I should have mentioned that this is a direct quote from Cask's very own FAQ (from its README.org)!

> It's all good though - I will just keep the restored static-if macro locally in transient.el for now. Thanks!

Please do this anyway:

> But first, if you haven't done so yet, then please post the output of tree -a elpa .cask first.

This would help me verify my suspicion and that in turn would help me help other Cask users.

**Reply by petergaultney** [reply]
I'm having this same problem and I don't use cask at all. I do use `straight.el` - maybe there's some similarity?

**Comment by starenka** [comment]
@mattsears `It's all good though - I will just keep the restored static-if macro locally in transient.el for now. Thanks!` thanks, it worked for me too

**Comment by julian-hoch** [comment]
Hi, I have the same issue but am somewhat lost as to how the resolution works. From what I understood there are two options:

a) downgrade transient package to an older version (how exactly did you achieve this? from my limited understanding of emacs installing older packages does not seem to be supported)
b) "locally restoring static-if macro in transient.el" - any tips on how to do this, exactly?

Right now, none of the transient-dependent packages work in my emacs instance, which is a bummer. This even happens with emacs -Q and a minimal setup, so it seems really tricky to resolve (short of starting from scratch, I guess).

**Reply by tarsius** [reply]
> From what I understood there are two options

These are not the approaches you should try. They worked for the other users in this thread, but I am 99% certain that this is only the case because that by doing what they did, they ended up unknowingly also doing the thing that actually fixes the issue, which is to recompile `transient.el` without either an older version of that package leaking in the new compilation result, or using a different version of Emacs for compilation than the one they use interactively.

Let's try to confirm that.

1. Do `M-: (locate-library "transient") RET`.  Tell me what the result is. This should end with `.elc`, if it doesn't, then do not proceed.
2. Exit Emacs.
3. Do `rm FILE-FROM-STEP-ONE`.
4. Start Emacs.
5. Try to reproduce the error. I suspect you won't be able to.

**Comment by julian-hoch** [comment]
Hi Tarsius,
thanks for getting back to me, I appreciate you trying to help!
I followed your steps, and found the .elc file in my "~/.config/emacs/straight/build/transient" directory. I deleted it and restarted emacs, but unfortunately it did not change the error.
I managed to previously remedy this by deleting transient package (although that was kind of tedious since I had to manually remove all dependencies first), but the next time I restarted emacs it reappeared.

**Reply by tarsius** [reply]
I think we can only figure out what is going wrong if someone who experiences this issue, very carefully analyses the situation. I cannot reproduce the issue, so it cannot be me.

For example, since apparently defining `static-if` in `transient.el` "fixes" the issue, you could investigate what is going on there. Which isn't the same as just putting that definition there and calling it a day. It makes no sense that doing that fixes the issue. I am not saying that doing that does not make the issue go away, just that correlation isn't causation. If you decide to investigate along those lines, you could, before adding that macro definition, check if that it is already defined and if so where. Maybe do that adding a debug statement to that effect to `transient.el` instead of the macro definition.

Given an error `transient--init-suffix-key is already defined as something else than a generic function`, it also makes sense to investigate as what it is defined and where. You might want to do so at different points in time. Before loading `transient` and then again right before that tries to define these methods, using a debug statement.

**Reply by julian-hoch** [reply]
Thanks for the tips. I am definitely willing to spend some time analysing and debugging this, since the alternative for me is probably to spend half a day completely rewriting my emacs config and potentially wiping my emacs configuration directory or something radical along these lines.
Unfortunately, my emacs lisp skills are not quite as advanced as I would like, but ChatGPT helped me with the following macro:

```emacs-lisp
    (defmacro my-debug-transient-key (&optional label)
      "Debug the definition of `transient--init-suffix-key`.
    Optional argument LABEL is used to label the debug output."
      `(let ((label (or ,label "Debug")))
         (message "%s: Checking `transient--init-suffix-key`..." label)
         (if (fboundp 'transient--init-suffix-key)
             (progn
               (message "%s: `transient--init-suffix-key` is defined as: %S" label (symbol-function 'transient--init-suffix-key))
               (with-output-to-temp-buffer "*transient-backtrace*"
                 (backtrace)))
           (message "%s: `transient--init-suffix-key` is not defined" label))))
```

That helped me figure out that the function switched from "not defined" to "dummy" right after I loaded package "rg", which matches with the following warning I also see in the **Warnings** buffer:

> ⛔ Error (use-package): rg/:catch: Invalid function: static-if

I commented out the "rg" package.  After that, transient loads fine!  So, I guess I will have to take this up with the "rg" package maintainers, or find an alternative search provider.

**Reply by tarsius** [reply]
Or maybe `rg` was just compiled a long time ago, with an older `eieio` and/or `transient`, not fully compatible with the current versions. So first simply reinstall `rg`.

**Reply by julian-hoch** [reply]
I tried that, however it did not help, however, simply moving `use-package` snippet for `rg` to the end of my config did help.
Certainly not the most elegant solution, but at least it solves it for ME. I think it's still a bit unclear WHY this is happening, but I guess Emacs sometimes works in mysterious ways 🤷

**Reply by tarsius** [reply]
"dummy" lead me to `cl-generic-define-method`. Looking through its history brought up https://github.com/emacsmirror/emacs/commit/ff053c706cb903534936c45b04a2aa38e3db1261, linking to https://debbugs.gnu.org/cgi/bugreport.cgi?bug=49053 and https://debbugs.gnu.org/cgi/bugreport.cgi?bug=47454.

Not sure what to make of that. Apparently that sort of thing could happen even within emacs itself and why that is so they didn't fully understand. These issues were closed with a tentative fix that made it into emacs 28.1.

**Reply by julian-hoch** [reply]
Wish I could help, but that's currently way above my paygrade...

**Reply by Gijs-Koot** [reply]
I ran into the same problem (`static-if` undefined and `transient` trying to reassign an already assigned function) using Spacemacs with Emacs 30.0 on WSL. In my case the error was triggered first at `journalctl`, but if I avoided loading that the errors showed up in another problem. I eventually "fixed" this for me by downgrading to Emacs 29.4. That is also a recompilation, so not sure if recompiling Emacs 30.0 would have also solved the problem. Recompiling or removing packages did not help.

**Reply by petergaultney** [reply]
@julian-hoch i know even less elisp than you do - I can't figure out how to _use_ your macro to debug this. Any tips?

**Reply by petergaultney** [reply]
@Gijs-Koot unfortunately i am on emacs 29.1 and still have the same issue. this happened to me after using `straight` to upgrade various packages.

**Reply by julian-hoch** [reply]
> @julian-hoch i know even less elisp than you do - I can't figure out how to _use_ your macro to debug this. Any tips?

Yeah, sure.  Copy the macro into your `*scratch*` buffer, and then evaluate it (e.g. using `C-x C-e` with the point at the end of the macro). Ideally put it also into your init.el so you dont have to repeat that when you restart emacs.

Then, go to your =init.el= file, and put the execution line (simply `(my-debug-transient-key "some-text-you-chose")`) at different points in the file - the easiest solution to be frank is to put the line at the beginning, somewhere in the middle, somewhere in the end (with different labels), and make sure it is not interrupting another top level expression.

Then, restart Emacs, and open your *Messages* buffer.  You should see the messages from the macro in there, and  the first ones will show 'undefined', the following ones will show 'dummy' (if memory serves me right).
You thus know more or less were the problem is (between the two lines where it changed).  To narrow it down further, introduce more debug lines in between the two lines, and so on.  Restart each time and check the *Messages* buffer.  Takes a while, but not too long - it's actually mathematically proven that you can find the error in a number of step that grows not faster than logarithmically with the number of lines in your file, which is pretty good as algorithms go.

**Reply by petergaultney** [reply]
thanks Julian! i tried something like that, and surprisingly, nothing ever said anything about this particular function getting installed at all. I should have tried adapting your macro to look at `static-if`, which was the new error i was getting, but @tarsius's push to `transient` today has fixed my issue immediately, so I'll just try to keep this snippet (and the new knowledge) in my back pocket for future confusing situations!

**Comment by tarsius** [comment, marked as answer]
I have pushed https://github.com/magit/transient/commit/0e35673ef5cfb4ae9ec8047a34d652b2f9986db6 which I hope will address this issue.
